File: leetcode/1_two_sum.java

/**
 * 1. Two Sum
 * Easy
 * Array, Hash Table
 * 
 * Map operations
 * 1. .containsKey()
 * 2. .get(<k>)
 * 3. .put(<k>, <v>)
 */

import java.util.Map;
import java.util.HashMap;

class Solution {
  public int[] twoSum(int[] nums, int target) {
    // Create hash map
    Map<Integer, Integer> prevMap = new HashMap<>();

    for (int i = 0; i < nums.length; i++) {
      int diff = target - nums[i];

      if (prevMap.containsKey(diff)) {
        return new int[] { prevMap.get(diff), i };
      }
      prevMap.put(nums[i], i);
    }
    return new int[0]; // Guranteed solution, no need for return
  }
}

File: leetcode/1_two_sum.py
"""
1. Two Sum
- Easy
- Array, Hash Table
- Link: https://leetcode.com/problems/two-sum/
"""

# Approach 1: One-Pass Hash Table
# Time: O(n) | Space: O(n)
class Solution:
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        dic = {}
        for i, num in enumerate(nums):
            n = target - num
            if n not in dic:
                dic[num] = i
            else:
                return [dic[n], i]
